File: cpp_service_tutorial/src/spawn_model_client.cpp
// gazebo model spawn by rosservice
// referenced from answers.ros.org
// url : https://pastebin.com/UTWJSScZ

// basic template of roscpp service client
// referenced from wiki.ros.org
// url : http://wiki.ros.org/ROS/Tutorials/WritingServiceClient%28c%2B%2B%29

#include <ros/ros.h>
#include <ros/package.h>
#include <gazebo_msgs/SpawnModel.h>
#include <geometry_msgs/Pose.h>
#include <iostream>
#include <fstream>

void addXml(gazebo_msgs::SpawnModel& model_in, const std::string& file_path ){
    std::ifstream file(file_path);
    std::string line;

    while (!file.eof()){
        std::getline(file, line);
        model_in.request.model_xml += line;
    }
    file.close();
}

geometry_msgs::Pose getPose(){
    geometry_msgs::Pose initial_pose;

    initial_pose.position.x = -2;
    initial_pose.position.y = 1;
    initial_pose.position.z = 1;

    initial_pose.orientation.z = -0.707;
    initial_pose.orientation.w = 0.707;

    return initial_pose;
}

int main(int argc, char** argv){

    ros::init(argc, argv, "gazebo_spawn_model");

    ros::NodeHandle nh;
    ros::ServiceClient spawn_model_prox = nh.serviceClient<gazebo_msgs::SpawnModel>("gazebo/spawn_urdf_model");
    
    gazebo_msgs::SpawnModel model;

    // add roslib in find_package()
    auto file_path = ros::package::getPath("cpp_service_tutorial") +  "/models/r2d2.urdf";

    addXml(model, file_path);

    model.request.model_name = "r2d2";
    model.request.reference_frame = "world";

    model.request.initial_pose = getPose();

    // ServiceClient.call() => return bool type
    if (spawn_model_prox.call(model)){
        auto response = model.response;
        ROS_INFO("%s", response.status_message.c_str()); // Print the result given by the service called
    }
    else {
        ROS_ERROR("Failed to call service /trajectory_by_name");
        return 1;   
    }

    return 0;
}

File: service_tutorial/src/spawn_model_client.py
#! /usr/bin/env python

"""
referenced from programcreek

url : https://www.programcreek.com/python/example/93572/rospkg.RosPack
"""

import rospy
import rospkg
from geometry_msgs.msg import Pose
from gazebo_msgs.srv import SpawnModel

rospy.init_node("gazebo_spawn_model")

# model_name
model_name = "r2d2"

# model_xml
rospack = rospkg.RosPack()
model_path = rospack.get_path("service_tutorial") + "/models/"

with open(model_path + model_name + ".urdf", "r") as xml_file:
    model_xml = xml_file.read().replace("\n", "")

# robot_namespace
robot_namespace = ""

# initial_pose
initial_pose = Pose()
initial_pose.position.x = -2
initial_pose.position.y = 1
initial_pose.position.z = 1

# z rotation -pi/2 to Quaternion
initial_pose.orientation.z = -0.707
initial_pose.orientation.w = 0.707

# reference_frame
reference_frame = "world"

# service call
spawn_model_prox = rospy.ServiceProxy("gazebo/spawn_urdf_model", SpawnModel)
result = spawn_model_prox(
    model_name, model_xml, robot_namespace, initial_pose, reference_frame
)

""" result fromat
bool success
string status_message
"""

print(result)
